### PR TITLE
[abc:iview:show] Added support for full series playlists - additional episode collection allows immediate checking against the archive

### PR DESCRIPTION
**Pull request by davidphood**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Added new extractor to support playlists for all episodes currently available for a given series. For example: https://iview.abc.net.au/show/play-school-story-time will download the full playlist of 24 episodes. The playlist entries are 'url' type and use the [abc:iview] extractor for the actual video download.

Minor changes have been made to the [abc:iview] extractor to support the url returned by the series api.

### CODE REVIEW

File: youtube_dl/extractor/abc.py [hunk 1]
```diff
@@ -191,3 +222,71 @@ def tokenize_url(url, token):
             'subtitles': subtitles,
             'is_live': is_live,
         }
+
+
+class ABCIViewSeriesIE(ABCIViewIE):
+    IE_NAME = 'abc.net.au:iview:series'
+    _VALID_URL = r'https?://iview\.abc\.net\.au/(show|programs)/(?P<id>[a-z0-9\-]+)/?$'
```
Comment (dstftw) [inline youtube_dl/extractor/abc.py]: No trailing `$`, override `suitable`.

File: youtube_dl/extractor/abc.py [hunk 2]
```diff
@@ -191,3 +222,71 @@ def tokenize_url(url, token):
...
+                'duration': int_or_none(episode.get('duration')),
+                'timestamp': parse_iso8601(episode.get('pubDate'), ' '),
+                'series': unescapeHTML(episode.get('seriesTitle')),
+                'series_id': episode.get('seriesHouseNumber') or video_id[:7],
+                'season_number': int_or_none(self._search_regex(
+                    r'\bSeries\s+(\d+)\b', title, 'season number', default=None)),
+                'episode_number': int_or_none(self._search_regex(
+                    r'\bEp\s+(\d+)\b', title, 'episode number', default=None)),
+                'episode_id': house_number,
+                'uploader_id': episode.get('channel'),
```
Comment (dstftw) [inline youtube_dl/extractor/abc.py]: Code duplication.

File: youtube_dl/extractor/abc.py [hunk 3]
```diff
@@ -191,3 +222,71 @@ def tokenize_url(url, token):
...
+            }
+            entries.append(entry)
+
+        return {
+            '_type': 'playlist',
+            'title': series_params.get('seriesTitle'),
+            'description': series_params.get('seriesDescription'),
+            'uploader_id': entries[0].get('uploader_id'),
```
Comment (dstftw) [inline youtube_dl/extractor/abc.py]: Breaks on empty `entries`.

File: youtube_dl/extractor/abc.py [hunk 4]
```diff
@@ -105,22 +105,53 @@ def _real_extract(self, url):
 
 class ABCIViewIE(InfoExtractor):
     IE_NAME = 'abc.net.au:iview'
-    _VALID_URL = r'https?://iview\.abc\.net\.au/(?:[^/]+/)*video/(?P<id>[^/?#]+)'
+    _VALID_URL = r'https?://iview\.abc\.net\.au/((?:[^/]+/)*video/|programs/(?:[^/]+/)*)(?P<id>[A-Z0-9]+)'
```
Comment (dstftw) [inline youtube_dl/extractor/abc.py]: 1. Do not restrict `id` expression.
2. Do not capture groups you don't use. Read coding conventions.

**Comment by section83** [comment]
This would be a great help to Australian users. However, by relying on the API, it might not pick up all related videos.  ABC iView have a number of related videos for many shows which are usually called "Extras". They can be behind the scenes videos or collections of deleted scenes etc.  As far as I can tell the URLs of those related videos are not available in the API pages (ie. iview.abc.net.au/api/series/[show-name]/).  I've not yet been able to locate a reliable source for those URLs, except the originating page where they are shown to users (ie. iview.abc.net.au/show/[show-name]/).

I'm in Oz and so not limited by geo-blocking.  I'd like to test these proposed changes but don't know how to DL a copy of the altered ytdl binary.  Can someone help ?

**Comment by deverton** [comment]
I've tested this patch locally from Australia and it does work fine. I'd consider downloading all the extras a separate feature.

**Comment by section83** [comment]
Thanks, glad to know it works.  I'm still hoping the changes can be implemented in the iView extractor.  In the meantime, I'm parsing the iView API pages to dig out URLs for episodes and extras.

**Comment by davidphood** [comment]
Hi @dstftw, I believe all requested changes have been resolved and the new version works as expected.

**Comment by deverton** [comment]
This patch is still working for me. Apart from fixing the conflict, what's missing?

**Comment by smlx** [comment]
Hey, this has been broken for a while. What's left to get this PR merged?

**Comment by dirkf** [comment]
Superseded by #26318

**Pull request closed without merging**